### PR TITLE
Feature/cets

### PR DESCRIPTION
**Pull request by chrzaszcz**
This PR checks if the CETS code from the feature branch passes the tests.
Ready for the final review and merge.

The following will be done after the merge:
- Add missing docs for some options, e.g. `component_backend`
- Update the migration guide with a description of `internal_databases` and the need to specify `mnesia` explicitly.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_25_arm64](https://circleci.com/gh/esl/MongooseIM/181316) / small_tests / e84d8169910f66736e756d7701f954dac5237770
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/181316/small_tests.25.3-arm64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/181316/small_tests.25.3-arm64/small/ct_run.mongooseim@localhost.2023-07-19_14.13.53/lib.mongooseim.logs/run.2023-07-19_14.13.53/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/181853) / small_tests / 5f98eeb01f6cf3ec3c911e2ec2bd22688d6df88b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/181853/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/181853/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-07-21_11.23.40/lib.mongooseim.logs/run.2023-07-21_11.23.40/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/183735) / small_tests / 34078499bd3577ad6ef970d9e1b5965a290e7a28
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/183735/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/183735/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-08-02_08.46.37/lib.mongooseim.logs/run.2023-08-02_08.46.37/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/184625) / small_tests / a3647acc646551c4c7667b10fdf9e435da8198cc
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/184625/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/184625/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-08-07_07.48.51/lib.mongooseim.logs/run.2023-08-07_07.48.52/suite.log.html)

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/184670) / elasticsearch_and_cassandra_mnesia / e2369988350da9a0333ae10242dc0e5ffa6d076d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/184670/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/184670/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40c2a000b5e63a.2023-08-07_08.04.54/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/184772) / elasticsearch_and_cassandra_mnesia / 411c00708fad9f653beb7317e53d9338f37f5748
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/184772/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/184772/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%40f7aeee6844af.2023-08-07_09.38.55/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/185285) / small_tests / 48b9569ee827cd859e7282ef7875ec0ef2bf0bd0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/185285/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/185285/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-08-08_07.12.42/lib.mongooseim.logs/run.2023-08-08_07.12.42/suite.log.html)

**Review by NelsonVides — CHANGES_REQUESTED** [review]
Ok, with such a gigantic diff, I'm not going to carefully read and verify everything changed here, some sections I'll only skim over, trusting that previous rounds of review and tests went well.
Notes:
- I've checked documentation, seems good to me. Not extensive enough but what is there is fine.
- I'm not verifying that everything mnesia now has a cets alternative, I trust having disabled mnesia in tests checked that already.
- I'm not paying thorough attention to pieces like bosh, jingle, or regular muc, I've only read the main module and it seems good, I trust their respective backends and/or helper modules are cool. Though btw, if we already created a dir for `muc` with the `_online_` helpers, we might have as well put `mod_muc` in there. Not important, just mentioning.
- Changes to ejabberd_router, components, and s2s, they're amazing, thank you!

-----
Some comments request small changes, specially the one in `ejabberd_app:start/2`, please fix that one 🙏🏽

File: src/mongoose_node_num.erl [hunk 1]
```diff
@@ -0,0 +1,30 @@
+%% Returns a numeric id from 0 to 255 for the current node.
+%% Used to generate MAM IDs.
+-module(mongoose_node_num).
+-export([set_node_num/1]).
+-export([node_num/0]).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+-include("mongoose_config_spec.hrl").
+-include("mongoose_logger.hrl").
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/mongoose_node_num.erl:10]: sssh, don't be Hank xd

File: rebar.config [hunk 2]
```diff
@@ -80,6 +80,7 @@
   {cache_tab, "1.0.30"},
   {segmented_cache, "0.3.0"},
   {worker_pool, "6.0.1"},
+  {cets, {git, "https://github.com/esl/cets.git", {branch, "main"}}},
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply rebar.config:83]: Yes, we will do before release (or you could try to make a package, hope the name is free... or we would have to use short eslcets)

File: src/ejabberd_app.erl [hunk 3]
```diff
@@ -40,18 +40,31 @@
...
+    application:start(cache_tab),
 
     mongoose_graphql:init(),
     translate:start(),
-    ejabberd_node_id:start(),
     ejabberd_commands:init(),
     mongoose_graphql_commands:start(),
+    mongoose_config:start(),
+    mongoose_metrics:init(),
+    db_init(),
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply src/ejabberd_app.erl:67]: This one I'll insist is strictly required, please fix.

File: src/ejabberd_sup.erl [hunk 4]
```diff
@@ -156,12 +156,19 @@ init([]) ->
     PG =
         {pg,
           {pg, start_link, [mim_scope]},
-          permanent, infinity, supervisor, [pg]},
+          permanent, infinity, worker, [pg]},
```
[the remark this answers is not in the document]
Comment (arcusfelis) [reply src/ejabberd_sup.erl:159]: rewrite it all, there is a lot of repetition in this module...

**Review by NelsonVides — APPROVED** [review]
LGTM 🔥🔥🔥

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_25](https://circleci.com/gh/esl/MongooseIM/185737) / elasticsearch_and_cassandra_mnesia / c5faa49326f12ab22e3444f857b9b8a126194a2e
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4063/185737/elasticsearch_and_cassandra_mnesia.25.3-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4063/185737/elasticsearch_and_cassandra_mnesia.25.3-amd64/big/ct_run.test%4001180d2f997e.2023-08-08_14.34.20/index.html)
**OK: 369** / Failed: 0 / **User-skipped: 38** / Auto-skipped: 0

**Pull request merged**